packages/features/tasker/repository.ts

```diff
@@ -0,0 +1,148 @@
+};
+
+const whereMaxAttempsReached: Prisma.TaskWhereInput = {
+  attempts: {
+    equals: {
+      // @ts-expect-error prisma is tripping: '_ref' does not exist in type 'FieldRef<"Task", "Int">'
+      _ref: "maxAttempts",
+      _container: "Task",
+    },
+  },
+};
+
+const whereUpcomingTasks: Prisma.TaskWhereInput = {
+  // Get only tasks that have not succeeded yet
+  succeededAt: null,
+  // Get only tasks that are scheduled to run now or in the past
+  scheduledAt: {
+    lt: new Date(),
+  },
+  // Get only tasks where maxAttemps has not been reached
+  attempts: {
+    lt: {
+      // @ts-expect-error prisma is tripping: '_ref' does not exist in type 'FieldRef<"Task", "Int">'
+      _ref: "maxAttempts",
+      _container: "Task",
+    },
+  },
+};
+
+export class Task {
+  static async create(
+    type: TaskTypes,
+    payload: string,
+    options: { scheduledAt?: Date; maxAttempts?: number } = {}
+  ) {
+    const { scheduledAt, maxAttempts } = options;
+    const newTask = await db.task.create({
+      data: {
+        payload,
+        type,
+        scheduledAt,
+        maxAttempts,
+      },
+    });
+    return newTask;
+  }
+  static async getNextBatch() {
+    const tasks = await db.task.findMany({
+      where: whereUpcomingTasks,
+      orderBy: {
+        scheduledAt: "asc",
+      },
+      take: 100,
+    });
+    return tasks;
+  }
+  static async getAll() {
+    const tasks = await db.task.findMany();
+    return tasks;
+  }
+  static async getFailed() {
+    const tasks = await db.task.findMany({
+      where: whereMaxAttempsReached,
+    });
+    return tasks;
+  }
+  static async getSucceeded() {
+    const tasks = await db.task.findMany({
+      where: whereSucceeded,
+    });
+    return tasks;
+  }
+  static async count() {
+    const tasks = await db.task.count();
+    return tasks;
+  }
+  static async countUpcoming() {
+    const tasks = await db.task.count({
+      where: whereUpcomingTasks,
+    });
+    return tasks;
+  }
+  static async countFailed() {
+    const tasks = await db.task.count({
+      where: whereMaxAttempsReached,
+    });
+    return tasks;
+  }
+  static async countSucceeded() {
+    const tasks = await db.task.count({
+      where: whereSucceeded,
+    });
+    return tasks;
+  }
+  static async retry(taskId: number, lastError?: string) {
+    const task = await db.task.update({
+      where: {
+        id: taskId,
+      },
+      data: {
+        attempts: { increment: 1 },
+        lastError,
+      },
+    });
+    return task;
+  }
+  static async succeed(taskId: number) {
+    const task = await db.task.update({
+      where: {
+        id: taskId,
+      },
+      data: {
+        attempts: { increment: 1 },
+        succeededAt: new Date(),
+      },
+    });
+    return task;
+  }
+  static async cancel(taskId: number) {
+    const task = await db.task.delete({
+      where: {
+        id: taskId,
+      },
+    });
+    return task;
+  }
+  static async cleanup() {
```

Comment:
I think we need to probably follow the repository pattern, if we're already abstracting this to be DB agnostic on the actual handling, could do the same as Orgs v2 does.

Reply:
I didn't want mix these patterns up. As the Task model is intended to be separate from the other models. Hence the lack of relationships.